lease.go

```diff
@@ -48,7 +50,36 @@ func (c *Client) WithLease(ctx context.Context, opts ...leases.Opt) (context.Con
 	}
 
 	ctx = leases.WithLease(ctx, l.ID)
-	return ctx, func(ctx context.Context) error {
-		return ls.Delete(ctx, l)
+	return ctx, func(octx context.Context) error {
+		// NOTE: If octx.Err() is nil, that means that octx is still
+		// working and we should return the error. If not, the cleanup
+		// caller might use canceled context to delete the temporary
+		// lease so that the lease will be leaky. The resources
+		// referenced by lease will be hold in 24 hours.
+		// Just in case, we should use brand new context to retry
+		// deletion.
+		//
+		// For example, the temporary content data hold by the leases
+		// increase disk pressure after cancel pulling image. Delete
+		// lease is good for the case.
+		//
+		// For the huge image, the temporary lease can hold the some
+		// downloaded layers and retry pulling can download whole image.
+		// But the deletion of leases might lead to endless of retry
+		// pulling. Anyway, good luck for this case!
+		err := ls.Delete(octx, l)
+		if err == nil || octx.Err() == nil {
```

Comment:
I've spent quite a while thinking and I still not sure I fully understand this line.
In which case we may end up with `err == nil && octx.Err() != nil` ?
If the `octx` errored, we always get a non nil `err`
If `err==nil`, then `octx.Err` is also nil, right? (unless we miss if check somewhere).
So basically `if err != nil` should be sufficient here?

Reply:
Sorry for late reply. The check condition will be:

* If `ls.Delete` return nil, it means that the temp lease has been deleted. So just return it.
* If `ls.Delete` return err, there are two cases:
    * there is something wrong from the lease manager, for example, not found error. For this case, the context isn't canceled (octx.Err() == nil ). We should not retry and just return.
    * If the context has been canceled, it maybe that the lease is still there and we can't delete it by the canceled context. Based on this, we should retry it and ignore the error.

Reply:
In our production, some pods use huge images and kubelet always canceled the ImagePull requests. Even if the kubelet uses backoff to handle retry, the expired time of lease is 24 hours. With multiple times of retry, the downloaded blob data has been hold by many more leases. If the host has disk pressure, the kubelet can't remove image actually because the blob is still hold by lease. For me, as operator, I need to login the host and remove the old leases. 

For my change, the deletion of the temporary lease makes the case worse: kubelet cannot pull the huge image successfully. 

The 24 hours is too long. Maybe we can change it to 2 hours?

Reply:
Seems like the issue is more related to using a canceled context here. We know that is an issue but right now leave it up to the caller. If we take this change, we might as well just always use a fresh context

Reply:
We can make it shorter, but this should only happen when the client complete goes away, if the client is still there tit is a bug to not cleanup.

Reply:
> We know that is an issue but right now leave it up to the caller. If we take this change, we might as well just always use a fresh context

If the WithLease isn't called by containerd library, the caller should handle cleanup. But the `Pull` client API can create tmp lease and it doesn't use fresh context. So it is good to make it shorter to prevent from adding fresh context anywhere.